### DISCUSSION TITLE
Target file resolve, How to force ignore artifacts from a specific site?

### DISCUSSION THREAD

**Discussion by jcompagner** [eclipse-tycho/tycho#332] [Q&A] [answered]
When i have this target file:

	    <location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
		    <repository location="https://download.eclipse.org/eclipse/updates/4.21/"/>
    	    <unit id="org.eclipse.equinox.executable.feature.group" version="3.8.1300.v20210828-0802"/>
		    <unit id="org.eclipse.sdk.ide" version="4.21.0.I20210906-0500"/>
	    </location>
	    <location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
		    <repository location="https://download.eclipse.org/releases/2021-09/"/>
            <unit id="org.eclipse.emf.sdk.feature.group" version="2.27.0.v20210816-1137"/>
            <unit id="org.eclipse.egit.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.http.apache.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.lfs.feature.group" version="5.13.0.202109080827-r"/>
    	    <unit id="org.eclipse.jst.web_ui.feature.feature.group" version="3.23.0.v202107101916"/>
	    </location>

and also besides that also our own p2 repo with some 3rd party plugins that are not shipped by eclipse/orbit

   <location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
            <repository location="https://developer.servoy.com/p2_repository/ship_plugins/2021.09/"/>
            <unit id="org.apache.aries.spifly.dynamic.bundle" version="1.3.3"/>

Then i now have a big problem because that aries.dynamic.bundle above has in the manifest imports like:

org.objectweb.asm;version="[9.1,10)",org.objectweb.asm.commons;version="[9.1,10)",org.objectweb.asm.util;version="[9.1,10)"

so it from 9.1 until 10.

but if you go to:

https://download.eclipse.org/releases/2021-09/202109151000/plugins/

you see that that update site also includes asm_9.2

which eclipse itself doesn't need or use:

https://download.eclipse.org/eclipse/updates/4.21/R-4.21-202109060500/plugins/

But i think because of that import tycho or p2 resolvement does try to match 9.2 that it finds in https://download.eclipse.org/releases/2021-09/

So i now get this error:

[ERROR] Internal error: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle sablo cannot be resolved:sablo [18]
[ERROR]   Unresolved requirement: Require-Bundle: slf4j.api
[ERROR]     -> Bundle-SymbolicName: slf4j.api; bundle-version="1.8.0.beta3-SNAPSHOT"
[ERROR]        slf4j.api [3]
[ERROR]          Unresolved requirement: Require-Capability: osgi.serviceloader; filter:="(osgi.serviceloader=org.slf4j.spi.SLF4JServiceProvider)"
[ERROR]            -> Provide-Capability: osgi.serviceloader; osgi.serviceloader="org.slf4j.spi.SLF4JServiceProvider"
[ERROR]               org.apache.logging.log4j.slf4j18-impl [11]
[ERROR]                 Unresolved requirement: Import-Package: org.slf4j.spi; version="[1.8.0,2.0.0)"
[ERROR]                   -> Export-Package: org.slf4j.spi; bundle-symbolic-name="slf4j.api"; bundle-version="1.8.0.beta3-SNAPSHOT"; version="1.8.0.beta3-SNAPSHOT"
[ERROR]                 Unresolved requirement: Require-Capability: osgi.extender; filter:="(osgi.extender=osgi.serviceloader.registrar)"
[ERROR]                   -> Provide-Capability: osgi.extender; osgi.extender="osgi.serviceloader.registrar"; version:Version="1.0.0"
[ERROR]                      org.apache.aries.spifly.dynamic.bundle [4]
[ERROR]                        Bundle was not resolved because of a uses constraint violation.
[ERROR]   org.apache.felix.resolver.reason.ReasonException: Uses constraint violation. Unable to resolve resource org.apache.aries.spifly.dynamic.bundle [osgi.identity; osgi.identity="org.apache.aries.spifly.dynamic.bundle"; type="osgi.bundle"; version:Version="1.3.3"] because it is exposed to package 'org.objectweb.asm' from resources org.objectweb.asm [osgi.identity; osgi.identity="org.objectweb.asm"; type="osgi.bundle"; version:Version="9.2.0.v20210813-1119"] and org.objectweb.asm [osgi.identity; osgi.identity="org.objectweb.asm"; type="osgi.bundle"; version:Version="9.1.0.v20210209-1849"] via two dependency chains.
[ERROR]
[ERROR] Chain 1:
[ERROR]   org.apache.aries.spifly.dynamic.bundle [osgi.identity; osgi.identity="org.apache.aries.spifly.dynamic.bundle"; type="osgi.bundle"; version:Version="1.3.3"]
[ERROR]     import: (&(osgi.wiring.package=org.objectweb.asm)(&(version>=9.1.0)(!(version>=10.0.0))))
[ERROR]      |
[ERROR]     export: osgi.wiring.package: org.objectweb.asm
[ERROR]   org.objectweb.asm [osgi.identity; osgi.identity="org.objectweb.asm"; type="osgi.bundle"; version:Version="9.2.0.v20210813-1119"]
[ERROR]
[ERROR] Chain 2:
[ERROR]   org.apache.aries.spifly.dynamic.bundle [osgi.identity; osgi.identity="org.apache.aries.spifly.dynamic.bundle"; type="osgi.bundle"; version:Version="1.3.3"]
[ERROR]     import: (&(osgi.wiring.package=org.objectweb.asm.commons)(&(version>=9.1.0)(!(version>=10.0.0))))
[ERROR]      |
[ERROR]     export: osgi.wiring.package=org.objectweb.asm.commons; uses:=org.objectweb.asm
[ERROR]   org.objectweb.asm.commons [osgi.identity; osgi.identity="org.objectweb.asm.commons"; type="osgi.bundle"; version:Version="9.1.0.v20210209-1849"]
[ERROR]     import: (&(osgi.wiring.package=org.objectweb.asm)(&(version>=9.1.0)(!(version>=9.2.0))))
[ERROR]      |
[ERROR]     export: osgi.wiring.package: org.objectweb.asm
[ERROR]   org.objectweb.asm [osgi.identity; osgi.identity="org.objectweb.asm"; type="osgi.bundle"; version:Version="9.1.0.v20210209-1849"]
              Unresolved requirement: Import-Package: org.slf4j; version="[1.8.0,2.0.0)"
[ERROR]                   -> Export-Package: org.slf4j; bundle-symbolic-name="slf4j.api"; bundle-version="1.8.0.beta3-SNAPSHOT"; version="1.8.0.beta3-SNAPSHOT"
[ERROR]                 Bundle was not resolved because of a uses constraint violation.
[ERROR]   org.apache.felix.resolver.reason.ReasonException: Uses constraint violation. Unable to resolve resource org.apache.aries.spifly.dynamic.bundle [osgi.identity; osgi.identity="org.apache.aries.spifly.dynamic.bundle"; type="osgi.bundle"; version:Version="1.3.3"] because it is exposed to package 'org.objectweb.asm' from resources org.objectweb.asm [osgi.identity; osgi.identity="org.objectweb.asm"; type="osgi.bundle"; version:Version="9.2.0.v20210813-1119"] and org.objectweb.asm [osgi.identity; osgi.identity="org.objectweb.asm"; type="osgi.bundle"; version:Version="9.1.0.v20210209-1849"] via two dependency chains.

because i think it finds 2 of them..

But it shouldn;t do that, i haven't selected those plugins, only some other features from https://download.eclipse.org/releases/2021-09/ like EGit.

How can i stil use the above urls and target file and force tycho/p2 to not use 9.2 for this but keep what the main Eclipse SDK just wants to use (9.1)?

I dont think this is fully tycho because also if i do an "export" from the target file editor and i see what it generates in the above scenario it also has both of those jar versions.

the only way that i can fix this at this moment is not to use:

	    <location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
		    <repository location="https://download.eclipse.org/releases/2021-09/"/>
            <unit id="org.eclipse.emf.sdk.feature.group" version="2.27.0.v20210816-1137"/>
            <unit id="org.eclipse.egit.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.http.apache.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.lfs.feature.group" version="5.13.0.202109080827-r"/>
    	    <unit id="org.eclipse.jst.web_ui.feature.feature.group" version="3.23.0.v202107101916"/>
	    </location>

but for all 3 features i import there use there own repo:

    <location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
            <repository location="https://download.eclipse.org/modeling/emf/emf/builds/release/2.27/"/>
            <unit id="org.eclipse.emf.sdk.feature.group" version="2.27.0.v20210816-1137"/>
        </location>
        <location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
            <repository location="https://download.eclipse.org/egit/updates-5.13/"/>
            <unit id="org.eclipse.egit.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.http.apache.feature.group" version="5.13.0.202109080827-r"/>
            <unit id="org.eclipse.jgit.lfs.feature.group" version="5.13.0.202109080827-r"/>
        </location>
        <location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
            <repository location="https://download.eclipse.org/webtools/downloads/drops/R3.23.0/R-3.23.0-20210822084517/repository/"/>
            <unit id="org.eclipse.jst.web_ui.feature.feature.group" version="3.23.0.v202107101916"/>
        </location>

if i do that, then it seems to be working again, the above error doesn't happen.
and also the target site that is created by using the export feature of the target file editor is also fine.

But that is a bit annoying that i now need to use 3 specific update sites instead of just 1

does a target file has an option to force exclude stuff?

**Comment by laeubi** [comment]
You can configure [exclusion filters](https://www.eclipse.org/tycho/sitedocs/target-platform-configuration/target-platform-configuration-mojo.html#filters), beside that Tycho 2.5.x has a possible fix for this kind of resolution errors.

**Reply by jcompagner** [reply]
i am using Tycho 2.5 for this already.
will try to exclusion filters, but it seems to me that P2 is more the problem (it also generates a p2 site where the wrong version jars are also added). I created also a case for that: https://bugs.eclipse.org/bugs/show_bug.cgi?id=576570

because to me it sounds that you really want this exclusion list inside the xx.target file..

**Reply by jcompagner** [reply]
now i think about this, i have to set this exclusion list over all the pom files that are using the my target file as the source..
Which is quite a few, and when eclipse moves itself to asm9.2 then i have to not forget to adjust all that again..

**Reply by laeubi** [reply]
That's strange, the use-constraint violations should not happen anymore with tycho 2.5.0 do you think you can derive a test-case from this?
We already have https://github.com/eclipse/tycho/tree/master/tycho-its/projects/resolver.usesConstraintViolations but this test passes so it would be interesting to find out what is different in your case. Target exclusions are currently not supported by tycho (and PDE has implemented it rather strange as a white-list instead of a black-list...)

**Reply by jcompagner** [reply]
let me try to make a copy of that and see if i can make it break

**Comment by laeubi** [comment]
> Then i now have a big problem because that aries.dynamic.bundle above has in the manifest imports like:

Just a note on this: Aries behaves correct here and having version ranges on package imports is actually a good thing...

**Reply by jcompagner** [reply]
yes i know, i think if they would do: org.objectweb.asm;version="[9.1,9.2)",
i wouldn't have my problem, but i guess they assume that this version should work for all minor releases, so i get why the have ,10)

**Comment by jcompagner** [comment, marked as answer]
ok after a lot of testing i finally did see that 1 pom file was still on tycho 2.3 ...
And changing that fixed at least the build problem, so tycho wise it can handle this now.

But i can't still use the target like that, because the product build that i do is just in my eyes not ok:

![image](https://user-images.githubusercontent.com/3274268/136971368-bf08bfe8-c167-4959-b513-45f96e4be5db.png)

all kinds of duplicates, no idea which one really comes up and does what..

if i use the target file that doesn't use the https://download.eclipse.org/releases/2021-09/ as a whole but takes the 3 things i get from there from there own sites, the generated product is nice

![image](https://user-images.githubusercontent.com/3274268/136972885-27e0d9fd-918b-46ca-a5e0-dbb8e0f0c533.png)


no doubles anymore, (asm itself and tree) and even utils is now just 9.1 and not only 9.2 (i wonder how that would match up at runtime)

**Comment by akurtakov** [comment]
I have a feeling that smth included in your product in 2021-09 have a requirement [9.2.0,..) which forces p2 to put both. Please check what else in your product requires asm and what are the version ranges.

**Reply by jcompagner** [reply]
no its really only that aries.dynamic.bundle that has  (1.3.3) that has [9.1,10)

Because if there was something that needed 9.2 why would it then build when it can't get to a 9.2 release? 
It builds fine with only 9.1 versions.. 

I think it really is because we have [9.1,10) and then P2 searches over all locations in the target if it can find one and it will find the 9.2 which is the highest so it also installs that. But the system (core eclipse) also needs 9.1 so that one is also installed.

In my eyes P2 shouldn't do that, P2 should only first search for jars on the installed UI's then it will find first 9.1 and that is fine and should just use that (don't search further) only when the installed jars of all the UI's that i selected don't have that jar or that jar version then search further.

**Reply by akurtakov** [reply]
I'm afraid this level of computation happens one level beneath p2 in  [sat4j](https://gitlab.ow2.org/sat4j/sat4j/) not finding the "optimal" solution although giving a "correct" one. If my suspicious is correct that might be very tough to fix as the project hasn't seen new releases in years.

**Reply by laeubi** [reply]
I'm not aware that P2 do any "global" optimization to reduce the number of dependencies (and this can even be undesirable).
 [9.1,10) means the greatest in the range from 9.1 excluding 10 in this case 9.2 is chosen. If any other dependency requires 9.1 (but can't use 9.2), this will also be included. If your target only contains 9.1, 9.1 will be used in all cases of course as it is a compatible choice, but given that semantic versioning is used 9.2 is *better* because it contains bugfixed and/or additional features!

**Reply by laeubi** [reply]
> not finding the "optimal" solution although giving a "correct" one

It is also not trivial to decide what is "optimal" (see my previous comment), as one could consider a minimal set of bundles optimal, while one might argue that more up-to-date bundles are more "optimal" ...

**Reply by jcompagner** [reply]
but the problem is ASM and that spifly is really core eclipse, that really is loaded very very early in all the stages
We had many problem with that already in the past (thats why we now just use a patched jface plugin, because byte code manipulation is horrible and very very tricky in OSGi)

Thats why i am very very careful that those asm things are not suddenly duplicate or weird, because i am already sure that that will cause problems in the future.  Its such core, it hooks up in to the weaving process and so on. (i have had many case at bugs.eclipse.org for that, some which couldn't just be fixed, hence the https://github.com/Servoy/servoy-eclipse/tree/master/org.eclipse.jface patch we now have to not use the weaving and bytecode manipulation things anymore)

We use spifly to hook up the logging jars (slf4j and the the bridge itself and log4j) into osgi through that service loader stuff.
By itself i dont think we use ASM, but it will be loaded really early because spifly does depend on it and if that then load 9.2 but then eclipse somewhere else loads 9.1 and maybe that is suddenly mixing (because it is so core and low level) then i am now already knowing at some point there will be a failure..

If it was just another kind of thing like commons.io or what ever. then i wouldn't be to bothered 
The only problem i can think of that also can happen ins the javax.servlet stuff, thats also tricky in our product (we run a tomcat internally and all kinds of plugins contribute servlets/filters to it) and you really need to be sure that all those plugins and tomcat uses the same classes else you get those weird javax.servlet.Servlet != javax.servlet.Servlet things. .

So yes i am very wary about duplicate jars, suddenly you need to be very careful, so i am in the "minimal set" camp.
If there are multiply request (imports) of the same jar with different version but they do overlap, make sure to ship only the jar that both are fine with. Only when the version requirements don't overlap you have to ship both.

**Reply by laeubi** [reply]
As told before you can explicitly filter out unwanted units if you really car, that's nothing P2 or tycho can decide for you, I can only explain *why* P2 chooses this set and you are getting different results, but that's nothing to blame spyfly also, you are simply running a very special setup with special demands and thus require special configurations ...

**Reply by jcompagner** [reply]
ok thats where we don't agree, this is not a very special configuration, i don't see that. Its for me very broad this can happen to any body building an eclipse product and doesn't expect that suddenly duplicate versions are suddenly loaded
Problems at runtime are very tricky to see and get to. I have in my team developers that spend hours maybe even days because the equals of something was just false which should be true because of the same class loaded twice from different jars...

**Reply by akurtakov** [reply]
So please read [this] (https://wiki.eclipse.org/Equinox_P2_Resolution#Size_of_the_problem_and_other_relevant_facts) . In short the goal when provisioning new is always to:
* Have highest version of each IU possible
* Have least number of IUs
If asm 9.2 had a single IU additionally required it would have been dropped and only 9.1 would have been in the solution. But they seem to be same which effectively leads to making a decision which of the two requirements has priority over the other one. 
So much with the history/theory side. Unfortunately, I don't have the knowledge about the code dealing with that to help out more.

**Reply by jcompagner** [reply]
by the way the problem with the filters is again that you need to define it where you use the target. (the target platform mojo) thats for me not so nice, that means adding that to 5 or 6 poms and maintaining that.

If i could specify the filter on the target itself that would be helpful then it would be in 1 place .

**Reply by laeubi** [reply]
> Ok thats where we don't agree, this is not a very special configuration, i don't see that.

OSGi is specifically designed to handle the case (very well) of multiple versions of a software-component running in the same VM. So if your product suffer from that in so many places this is 'special' in the sense that the normal case is that it dosen't mater much (regarding correctness and function) otherwise.

> eclipse product and doesn't expect that suddenly duplicate versions are suddenly loaded

There is nothing *suddenly* happening, you have just made your target that broadly and using component with different demands that you are now getting a non-unique solution regarding the bundle-symbolic-name but that is nothing to worry about or be concerned from the OSGi perspective, its just Eclipse that has some special demands (e.g. 'singleton bundles').

> Problems at runtime are very tricky to see and get to

If you know how OSGi works and use the right tools its not tricky. It just became tricky because most people don't care how OSGi works and just *want that it works* regardless of what would be the correct approach and doing dirty class-loader tricks instead of fixing the architecture or design but that's a different story.

**Reply by laeubi** [reply]
> If i could specify the filter on the target itself that would be helpful then it would be in 1 place .

You could place the target configuration in a parent pom and reference that in your projects.

**Reply by laeubi** [reply]
> So much with the history/theory side. Unfortunately, I don't have the knowledge about the code dealing with that to help out more.

The 'problem' is that the product install is sometimes incremental, so P2 has not always a chance to find a minimal solution. Also here 9.1 seems some hard requirement from some IU (don't know if P2 could emit info what ones), and package-import implies the highest version (9.2) (as per rule 1 from your comment) thus both versions are included.

OSGi can still use that only 9.1 is used if any usage-constraints require this and the 9.2 version will simply be not used and does nothing.

**Reply by jcompagner** [reply]
> > If i could specify the filter on the target itself that would be helpful then it would be in 1 place .
> 
> You could place the target configuration in a parent pom and reference that in your projects.

i could investigate that a bit, we have 8 git repo's that all supply a set of plugins that in the end will be a final product build
Those should all be using the same "target" 
But a parent pom is in this scenario not so nice because that mostly means also a directory structure.. (but maybe that can be worked around)

But thats besides this discussion anyway

**Reply by laeubi** [reply]
A parent pom can be deployed on nexus and be used without an explicit directory structure then.

**Reply by jcompagner** [reply]
> If you know how OSGi works and use the right tools its not tricky. It just became tricky because most people don't care how OSGi works and just _want that it works_ regardless of what would be the correct approach and doing dirty class-loader tricks instead of fixing the architecture or design but that's a different story.

i know how it works,
but as an example then how tricky it is

https://github.com/Servoy/servoy-eclipse-tomcat/blob/master/org.apache.tomcat/META-INF/MANIFEST.MF#L42

we have that project, that is an embedded tomcat inside our product 

besides that that is kind of a plain copy of tomcat that eclipse plugin also brings a extension point:

https://github.com/Servoy/servoy-eclipse-tomcat/blob/master/org.apache.tomcat/schema/serviceprovider.exsd

so any other plugin can just create filters or servlets and those are contributed to the root webapp that tomcat will deploy

like this project in a different git repo:

https://github.com/Servoy/servoy-eclipse/blob/master/com.servoy.eclipse.ngclient.ui/src/com/servoy/eclipse/ngclient/ui/ServicesProvider.java

now if tomcat that depends on the javax.servlet api jar version A gets through that ServicesProvider extension point a class pushed (filter or servlet impl) that is by that plugin depending on a slightly different javax.servlet api jar it would bomb out...
it woudn't load it because internally in tomcat something goes wrong because equals and that kind of stuff are not from classes he expects

so in this scenario, eclipse/osgi should not just have for 1 a minor release A and for the other a minor release A+1 
it must be the same thing.

pure looking at OSGi you would say it doesn't matter there are just 2 plugins that are loading 2 different sets of plugins but at runtime this is not working.
So yes all the imports (package or plugins) must be really correct to force eclipse to use the correct stuff, one mismatch and it will fail. And i think there are already a few different versions of stuff in the base of eclipse (servlet of X jsp of Y)

**Reply by laeubi** [reply]
> pure looking at OSGi you would say it doesn't matter there are just 2 plugins that are loading 2 different sets of plugins but at runtime this is not working.

It is working if you define your metadata correctly (e.g. export packages with proper uses constraints, import package with proper version ranges and possible attribute filters) because OSGi never resolves an incompatible class-space if you do not allow this. Thats why I said, people often simply don't care (e.g. use require bundle for no good reason, import packages without proper version constraint, use dirty classloading tricks) and then it can result in some diffuse problems.

But the solution is not to force *one* version for all but having proper meta-data instead... To keep at your example just take a look at:
https://github.com/Servoy/servoy-eclipse/blob/master/com.servoy.eclipse.ngclient.ui/META-INF/MANIFEST.MF

1. you require `org.eclipse.osgi` what almost always is problematic as this is the system bundle so your are have a hard requirement to the equinox framework and wire almost everything it exports to your bundle, better import the OSGi packages with consumer version range
2. same for org.apache.commons.io (I assume you won't need *everything* from it nor does it contains any split-packages)
3. You export `com.servoy.eclipse.ngclient.ui` but without declaring use-contraints on javax.servlet.* but seem to use it there, so you should let PDE calculate uses attributes to make sure the OSGi resolver only wires importers of `com.servoy.eclipse.ngclient.ui` to the same  `javax.servlet` packages. The same applies to your package exports of the Tomcat bundle.

Beside that, you might want to take a look at [pax-web](https://github.com/ops4j/org.ops4j.pax.web) what implements the [Web Applications Specification](https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.war.html) and [Http Whiteboard Specification](https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.http.whiteboard.html) that is allows OSGi compatible web technologies without eclipse specific extension points (that are not dynamic aware enough in most cases).

**Reply by jcompagner** [reply]
> 1. you require `org.eclipse.osgi` what almost always is problematic as this is the system bundle so your are have a hard requirement to the equinox framework and wire almost everything it exports to your bundle, better import the OSGi packages with consumer version range
> 2. same for org.apache.commons.io (I assume you won't need _everything_ from it nor does it contains any split-packages)

thx, yes i was planning to move stuff more from bundle imports to package imports (its already ungoing) 
but yeah X number of years ago it you would normally do the bundle import so that stays a bit.
But i know its better now to just do package imports (with a version range)

i did find that the hard way also in some bundles because there are currently bundles that have a different bundle name but they export the same packages..  (as an example every webcontainer (like tomcat) has there own javax.servlet bundle with ofcourse the same packages but they all name it differently)


> 3. You export `com.servoy.eclipse.ngclient.ui` but without declaring use-contraints on javax.servlet.* but seem to use it there, so you should let PDE calculate uses attributes to make sure the OSGi resolver only wires importers of `com.servoy.eclipse.ngclient.ui` to the same  `javax.servlet` packages. The same applies to your package exports of the Tomcat bundle.

thx, will have a look for this.

> 
> Beside that, you might want to take a look at [pax-web](https://github.com/ops4j/org.ops4j.pax.web) what implements the [Web Applications Specification](https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.war.html) and [Http Whiteboard Specification](https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.http.whiteboard.html) that is allows OSGi compatible web technologies without eclipse specific extension points (that are not dynamic aware enough in most cases).

our product is purely an eclipse product and we are in full control about that, so i don't care to much about that we use the eclipse extension point for this. Its not that that will be used in another setting.
But if the above is fixing also the case that we solve with the extension point that we  want to contribute @WebFilter or @WebServlet that is in developer just in other plugins, so not something tomcat will find in its context (so the tomcat scanner will find them), then i will have a look.
Because that is currently the only thing we solve, our product can also do an export (that generates a WAR) that then has all the Filters/Servlets in the WAR/WEB-INF/xxx so tomcat will find them. But when starting in our IDE, tomcat also needs them (or in some cases a special version of the filter that understands that it runs in an eclipse product (can access the workspace and so on)), so in this scenario we just give them programmatically and add them to the (root) tomcat context programmatically.

**Reply by laeubi** [reply]
With WABs (Web application bundles) you can also deploy "war-style" code but I have never used that...

**Reply by jcompagner** [reply]
i have to say the not so nice thing about package imports is this scenario:

public class CopySourceFolderAction extends Action

that Action is org.eclipse.jface.action

so if i removed the jface bundle then i get a compile error
i can fix that by having a quick fix on Action and import then that package (org.eclipse.jface.action)
But that doesn't fix the full problem because then you get:

"The hierarchy of the type CopySourceFolderAction is inconsistent"

which is not that easy to fix, you really need to go over the hierarchy and guess what you need to import by looking at interfaces or the extends classes..... (and happily we have the source....)

so it is not so nice that i also need to import the package: org.eclipse.core.commands.common
because Action extends AbstractAction which extends org.eclipse.core.commands.common.EventManager

**Reply by jcompagner** [reply]
ha ;)
https://bugs.eclipse.org/bugs/show_bug.cgi?id=571338#c3

I am not the only one that is complaining about that :)

**Reply by laeubi** [reply]
you - are - not- ALOOONNNNEEE ;-)

jface is probably a good candidate for require-bundle as most eclipse-ui bundles :-\
